src/fabric.c

```diff
@@ -507,6 +655,10 @@ void DEFAULT_SYMVER_PRE(fi_freeinfo)(struct fi_info *info)
 			free(info->fabric_attr->prov_name);
 			free(info->fabric_attr);
 		}
+		if (info->handle && (info->handle->fclass == FI_CLASS_NIC) &&
+		    info->handle->ops && info->handle->ops->close)
+			fi_close(info->handle);
+
```

Comment:
This change does not work with existing applications.  We cannot safely deference info->handle without the risk of accessing freed memory.

For MSG EPs, info->handle is used to report a connection request (fid FI_CLASS_CONNREQ).  That fid may be freed when an app calls fi_accept/fi_reject, but info->handle is left unmodified.

We need an alternate solution for obtaining/freeing the nic attributes.  :/

Reply:
Is `info->handle` overloaded with multiple values solely to avoid adding another member to the struct / for ABI reasons?

Reply:
Ok.  Since `handle` can be one of several different types, then it seems like it is required to also set `handle` back to NULL once the thing pointed to it is freed.